### PR TITLE
refactor(storage): store vec sstable info in staging and loosen batch id order check

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Currently, our `StagingSstableInfo` only stores a single sstable info. However, each `StagingSstableInfo` is generated from a local compaction task, which generates a vector of sstable info. Therefore, in this PR, we store a vector of sstable info instead.

Besides, we have loosen the batch id order check when we apply the staging sstable info. The reason is, first, generating imm and writing imm to staging sstable info is not atomic, and a smaller batch id may be written later than a greater one in corner case, and second, currently we share a single read version globally, and sometimes, without breaking the data consistency, some newer imms can be flushed and replaced by sstable earlier than some older imms.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/storage/src/hummock/store/version.rs [hunk 1]
```diff
@@ -82,9 +108,15 @@ impl StagingVersion {
             .iter()
             .filter(move |staging_sst| {
                 *staging_sst.epochs.last().expect("epochs not empty") <= epoch
-                    && filter_single_sst(&staging_sst.sst_info, table_id, key_range)
             })
-            .map(|staging_sst| &staging_sst.sst_info);
+            .flat_map(move |staging_sst| {
```
Comment (Li0k) [inline src/storage/src/hummock/store/version.rs:112]: need a related test case for multi state_table_infos for checking `prune_overlap`

**Review by hzxa21 — COMMENTED** [review]
LGTM.

File: src/storage/src/hummock/store/version.rs (hunk 1 above)
Comment (wenym1) [reply src/storage/src/hummock/store/version.rs:113]: I have expand our current unit test to cover the newly added code.

File: src/storage/src/hummock/store/version.rs [hunk 2]
```diff
@@ -99,6 +99,7 @@ impl StagingVersion {
         impl Iterator<Item = &SstableInfo> + 'a,
     ) {
         let overlapped_imms = self.imm.iter().filter(move |imm| {
+            println!("key_range: {:?}, start key: {:?}, end_key: {:?}, {:?}, {:?}", key_range, imm.start_user_key(), imm.end_user_key(), imm.start_key(), imm.end_key());
```
Comment (Li0k) [inline src/storage/src/hummock/store/version.rs:102]: remove `println!`

**Review by Li0k — APPROVED** [review]

**Pull request merged**